### PR TITLE
Create ns_speex_tilde

### PR DESCRIPTION
**Pull request by michalsoloducha**
I have tested it for different input parameters and it seems to work fine. Feel free to add modifications/comments.

### CODE REVIEW

File: src/signal-processing/ns_speex_tilde [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+Outlets:
+  1x Audio outlet
+
+*/
+
+#include <m_pd.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <string.h>
+#include "ringbuffer.h"
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Should not be required.
I will cleanup this include everywhere else.

File: src/signal-processing/ns_speex_tilde [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  t_float float_inlet;	// inlet (used for incoming audio)
+  t_outlet *audio_out;	// outlet
+
+  t_generic_codec codec;
+
+  void *speex_preprocess_state;
+
+} t_ns_speex_tilde;
+
+
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Remove empty line

File: src/signal-processing/ns_speex_tilde [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+// processing tree
+void ns_speex_tilde_dsp (t_ns_speex_tilde * x, t_signal ** sp) {
+  if (x->speex_preprocess_state != NULL) {
+    speex_preprocess_state_destroy (x->speex_preprocess_state);
+  }
+
+  x->speex_preprocess_state = speex_preprocess_state_init (x->codec.frame_size, x->codec.sample_rate_internal);
+
+  // enable noise suppression
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Remove comment

Comment (dennisguse) [reply src/signal-processing/ns_speex_tilde]: And also additional comments, e.g. constructor, destructure etc.

File: src/signal-processing/ns_speex_tilde [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  x->audio_out = outlet_new (&x->x_obj, &s_signal);
+
+  post ("ns_speex~: Created with frame size (%d), sampling rate (%f) and max. noise attenuation (%d).", x->codec.frame_size, x->codec.sample_rate_internal, x->max_noise_att);
+
+  return (void *) x;
+}
+
+// setup new class
+void ns_speex_tilde_setup (void) {
+  ns_speex_tilde_class = class_new (
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Please apply the gnu intend tool as described in the README.md

File: src/signal-processing/ns_speex_tilde [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	gensym ("ns_speex~"), 
+	(t_newmethod) ns_speex_tilde_new, 
+	(t_method) ns_speex_tilde_free, 
+	sizeof(t_ns_speex_tilde), 
+	CLASS_DEFAULT,
+	A_DEFFLOAT, A_DEFFLOAT, A_DEFFLOAT, 0 // init. values for: frame size, sample rate, max. noise attenuation
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Remove comment

File: src/signal-processing/ns_speex_tilde [hunk 6]
```diff
@@ -0,0 +1,167 @@
+/**
+@file ns_speex_tilde.c
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: Can you rename it to denoise_speex_tilde.c
This allows to understand easier what it actually does/

File: src/signal-processing/ns_speex_tilde [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+@date 2016-10-21
+@license GPLv3 or later
+
+ns_speex~ applies the _noise suppression_ algorithm of [Speex](http://www.speex.org/).
+
+Parameters:
+  ns_speex~ FRAME_SIZE SAMPLE_RATE MAX_NOISE_ATT
+  FRAME_SIZE: 80, 160, 240, 320, 640 [samples]
+  SAMPLE_RATE: 8000, 16000, 32000 [Hz]
+  MAX_NOISE_ATT - maximum noise attenuation (negative number) [dB]
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: @MAX_NOISE_ATT

I would prefer the long version (MAX_NOISE_ATTENUATION) and also the expected range -100..0?
What is the default value?

File: src/signal-processing/ns_speex_tilde [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+#include "speex/speex.h"
+#include "speex/speex_preprocess.h"
+
+static t_class *ns_speex_tilde_class;
+
+typedef struct _ns_speex_tilde {
+  t_object x_obj;
+  
+  t_int max_noise_att;
```
Comment (dennisguse) [inline src/signal-processing/ns_speex_tilde]: long version

**Comment by dennisguse** [comment]
Thanks Michal!
Its a good job!

I just added some comments, and when resolved I will merge it.
And could you add a help file?
I just saw that vad_speex~ also misses the help file and it would be awesome if you could also add one for this external.

**Review by dennisguse — CHANGES_REQUESTED** [review]
Looks good - just requires some smaller changes.

File: src/signal-processing/ns_speex_tilde (hunk 4 above)
Comment (michalsoloducha) [reply src/signal-processing/ns_speex_tilde]: I can't find that information in the main readme file.

Comment (dennisguse) [reply src/signal-processing/ns_speex_tilde]: Ohh my fault - it is in the DEVELOPER.md

```
indent \
--braces-after-if-line \
--braces-after-func-def-line \
--braces-on-if-line \
--braces-on-func-def-line \
--braces-on-struct-decl-line \
--break-after-boolean-operator \
--break-before-boolean-operator \
--dont-break-function-decl-args \
--dont-break-function-decl-args-end \
--dont-break-procedure-type \
--dont-star-comments \
--space-after-cast \
--no-tabs \
--cuddle-else \
-l999 \
*.{c,h}
```

**Comment by michalsoloducha** [comment]
Allright, all the requested changes are done. At some point later I will provide the help files for denoise and vad.

**Comment by dennisguse** [comment]
Code looks good.
I would like to wait until the help file is also ready.
Then it is just one commit ;)

**Comment by michalsoloducha** [comment]
I have uploaded the help files for denoise_speex and vad_speex. Plus some minor changes.

**Review by dennisguse — APPROVED** [review]

**Comment by dennisguse** [comment]
One final thing: is t_outlet actually used by denoise_speex_tilde.c?
The required outlet should be made available by generic_codec.h, or?

Could you test if t_outlet is required, and if not remove it?

**Comment by michalsoloducha** [comment]
Ok, so I got rid of unnecessary audio outlet declaration made some minor fixes.

**Pull request merged**